### PR TITLE
Support groups & boosting on `dynamic/slow`

### PR DESCRIPTION
**Pull request by OllysCoding**
## What does this change?

- Refactors `dynamic/slow` to use existing pattern for dynamic containers
- Add support for all layout types, supporting boosting & groups
- Creates a shared library for dynamic slices 'dynamicSlices.tsx'
- Creates 'shouldPadWrappableRows' abstraction for awkward padding logic
- Documents dynamic/slow variants in storybook

Related PRs
 - Refactor to same pattern for `dynamic/package`: https://github.com/guardian/dotcom-rendering/pull/5700
 - Introduction of this pattern for `dynamic/slow-mpu`: https://github.com/guardian/dotcom-rendering/pull/5617

## `dynamic/slow` overview

`dynamic/slow` supports the same grouping as `dynamic/fast` - huge, very big, big and standard. 

The container can be broken into two slices:

### First slice

The first slice is made up of either 1 huge, 1 very big or 2 very bigs. If there are 2 very bigs, either can be boosted, providing the following layouts:

| Name      | Screenshot      |
|-------------|------------|
| 1 huge | ![1huge][] |
| 1 very big | ![1verybig][] |
| 2 very bigs | ![2verybigs][] |
| 2 very bigs (1st boosted) | ![1stboosted][] |
| 2 very bigs (2nd boosted) | ![2ndboosted][] |

[1huge]: https://user-images.githubusercontent.com/9575458/187663680-aabe4d82-7c8b-47f2-9b65-048b00a04552.png
[1verybig]: https://user-images.githubusercontent.com/9575458/187663745-827afdaa-08dd-439e-8ae8-a313cc94fc64.png
[2verybigs]: https://user-images.githubusercontent.com/9575458/187663828-e70a57ff-4d36-473d-a06a-f60136adc1f1.png
[1stboosted]: https://user-images.githubusercontent.com/9575458/187663897-fdfa99af-9c09-46b7-91e1-dbde4865209e.png
[2ndboosted]: https://user-images.githubusercontent.com/9575458/187663960-ea164f85-a136-4016-b81f-2e7bb24c9c19.png

Notably, this is identical to the first slice in `dynamic/fast`, and the 2 very bigs + boosted variants are also used in `dynamic/slow-mpu`. Further refactors can help bring all of this logic & layouts in sync. 

### Second Slice

The second slice takes any remaining cards from the huge & very big groups that weren't used in the first slice, and demotes them to 'bigs', which is then used to decide the layouts for the second slice:

| Name      | Screenshot      |
|-------------|------------|
| 0 bigs | ![0bigs][] |
| 1 big | ![1big][] |
| 2 bigs | ![2bigs][] |

[0bigs]: https://user-images.githubusercontent.com/9575458/187674619-3526b17c-de03-44fa-8eef-c3bb5e56b134.png
[1big]: https://user-images.githubusercontent.com/9575458/187674682-4eb83c9b-6fcb-4f67-af56-86bd40e5253a.png
[2bigs]: https://user-images.githubusercontent.com/9575458/187674735-a9bff829-ecb5-46ff-997a-819898663adb.png

## Why?

1 step closer to parity :D

## Screenshots

As an example - 1 huge, 1 big, 4 standards
| Frontend      | DCR Before      | DCR After |
|-------------|------------|------------|
| ![frontend][] | ![dcrbefore][] | ![dcrafter][] |

[frontend]: https://user-images.githubusercontent.com/9575458/187675405-070ec51c-be1b-4297-b585-e53fd03f65d5.png
[dcrbefore]: https://user-images.githubusercontent.com/9575458/187675579-dd4f2db4-b6c1-43ba-8cb2-6dbb23b9ed64.png

[dcrafter]: https://user-images.githubusercontent.com/9575458/187675453-6eb0a9c8-fc3f-4d95-96a3-a9ff6ac66d8f.png

### CODE REVIEW

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * @param index - Index of the current card
+ * @param totalCards - Total number of cards being shown
+ * @param cardsPerRow - No. of cards in each row (if full)
+ */
+export const shouldPadWrappableRows = (
+	index: number,
+	totalCards: number,
+	cardsPerRow: number,
+): boolean => index < totalCards - (totalCards % cardsPerRow || cardsPerRow);
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:162]: Do you have an example of what you mean sorry? `gap` in terms of CSS grid?

File: dotcom-rendering/src/web/components/DynamicSlow.tsx [hunk 2]
```diff
@@ -1,126 +1,412 @@
...
+	const bigs = [
+		// Demote any left over 'huge' or 'veryBig' grouped cards
+		...secondSliceGroupedTrails.huge,
+		...secondSliceGroupedTrails.veryBig,
+		...secondSliceGroupedTrails.big,
+	];
+
+	// Decide the contents of the second slice
+	switch (bigs.length) {
+		case 0: {
```
Comment (oliverlloyd) [inline dotcom-rendering/src/web/components/DynamicSlow.tsx:302]: So simple

**Review by oliverlloyd — APPROVED** [review]
This is _really_ great. The business logic here is complex but I find this code easy to read through and understand, very impressive!

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		</LI>
+	</UL>
+);
+
+/* ._________________________._________.
+ * |#########|###############          |
+ * |#########|###############          |
+ * |_________|###############__________|
+ */
+export const Card25_Card75 = ({
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:103]: Yup :)

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
+/* eslint-disable @typescript-eslint/naming-convention -- because underscores work here*/
+import type { DCRContainerPalette } from '../../types/front';
+import type { TrailType } from '../../types/trails';
+import { LI } from '../components/Card/components/LI';
+import { UL } from '../components/Card/components/UL';
+import { FrontCard } from '../components/FrontCard';
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:6]: :D

File: dotcom-rendering/src/web/components/DynamicSlow.tsx [hunk 5]
```diff
@@ -1,126 +1,412 @@
...
+	containerPalette,
+	showAge,
+}: Props) => {
+	let firstSliceLayout:
+		| undefined // If there are no very bigs or huges, there is no first slice
+		| 'oneHuge'
+		| 'oneVeryBig'
+		| 'twoVeryBigs'
+		| 'TwoVeryBigsFirstBoosted'
+		| 'TwoVeryBigsSecondBoosted' = undefined;
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicSlow.tsx]: Fixed - tested and it wasn't necessary :)

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:162]: `gap` would be a massive improvement to how LI & UL work - will be very happy to delete this abstraction if we can get it working!

File: dotcom-rendering/src/web/components/DynamicSlow.tsx [hunk 6]
```diff
@@ -1,126 +1,412 @@
...
+				padSides={true}
+				showDivider={true}
+				padBottomOnMobile={true}
+			>
+				<FrontCard
+					trail={big}
+					containerPalette={containerPalette}
+					headlineSizeOnMobile="large"
+					showAge={showAge}
+					trailText={big.trailText}
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicSlow.tsx:103]: Good catch - upon investigating there are a few parts of this PR where sublinks / trail texts don't line up with where they should, so just working on a larger fix now

File: dotcom-rendering/src/web/components/DynamicSlow.tsx [hunk 7]
```diff
@@ -1,126 +1,412 @@
...
+						percentage="25%"
+						padSides={true}
+						padBottomOnMobile={true}
+						showDivider={true}
+					>
+						<FrontCard
+							trail={big}
+							containerPalette={containerPalette}
+							showAge={showAge}
+							trailText={big.trailText}
```
[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicSlow.tsx:164]: According to Frontend - we do, but if there are sublinks we should stop showing the trail text (coming in the next PR)

Confusingly, in the _oneHuge_ layout, we show neither, in _oneVeryBig_ we show only trail text, in _oneBig_ we show no trail text but do show sublinks, and in _twoBigs_ we show trail text only if there are no sublinks .... 

A larger refactor may be needed but I wonder if it's worth trying to refactor it into some kind of `additionalContent` config, e.g 

```js
additionalContentConfig={{
    trailText: true
    supportingContent: true
    supportingContentLayout: 'horizontal'
    hideTrailTextWhenSupportingContent: false
}} 
```

[the remark this answers is not in the document]
Comment (OllysCoding) [reply dotcom-rendering/src/web/components/DynamicSlow.tsx:164]: One of the things I find difficult now is that it's easy for it all to blend in together, but given the weirdness and inconistentness of fronts, I don't know whether we can pull this logic into Card.tsx and out of the containers

File: dotcom-rendering/src/web/lib/dynamicSlices.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (mxdvl) [reply dotcom-rendering/src/web/lib/dynamicSlices.tsx:162]: Yes, I thought it was going to be a really easy fix, but it’s a little bit more invovled. I’m working through the last quirks, but definitely not a blocker!

**Pull request merged**